_posts/2020-10-27-best-ec2-instance-for-rails-apps.md

```diff
@@ -0,0 +1,155 @@
+---
+layout: post
+title: "What's the Best EC2 Instance Type for Rails Apps?"
+date: 2020-10-27 10:00:00
+categories: ["rails", "performance", "ruby"]
+author: "noahgibbs"
+---
+
+<img src="/blog/assets/images/ec2_instance_pricing.png" style="max-width: 400px; float:right;" alt="EC2 us-east-1 on-demand instance prices - a huge grid of annoying numbers." />
+
+Do you ever [look at the list of Amazon EC2 instance types?](https://aws.amazon.com/ec2/instance-types). Those are sizes of virtual machine you can rent to run your code on. Well, okay, they're <i>groups of sizes</i>, since each one of those headings has a bunch of different sizes of VM...
+
+So ***what type of EC2 instances should you run your Rails app on***?
+
+The answer is simpler than it looks.
+
+Do you love numbers? I love numbers. Do you hate numbers? Skip to the bottom, there's a nice summary paragraph. Do you ***really really*** love numbers? There are [raw data dumps](https://codefol.io/links/ec2_inst_types_full_data.tar.bz2) including all my intermediate results.
+
+<!--more-->
+
+## Instance Types and How I'm Testing
+
+As you [may already know](https://www.fastruby.io/blog/rails/performance/ruby/hows-the-performance-of-ruby-3.0.0-preview1.html), I frequently run a "big Rails app, see how fast it can process many requests in parallel" benchmark called [Rails Ruby Bench](https://github.com/noahgibbs/rails_ruby_bench) (aka RRB.)
+
+First I'll run through what EC2 types to consider. Then I'll speed test the best choices.
+
+A Rails app is surprisingly CPU-bound. That's part of the reason we've seen a [72% speed increase from Ruby 2.0 to 2.6](https://engineering.appfolio.com/appfolio-engineering/2019/3/7/ruby-speed-roundup-20-through-26). That means that burstable-CPU instances such as t4 can sometimes show fantastic results on short benchmarks but worse results on longer-running benchmarks. If your app has extremely busy and extremely idle times this could be worth considering. But I can't provide reliable numbers for you for your app. It just depends how even the workload is. And the more even the workload is, the worse burstable instances are likely to do.
```

Comment:
@noahgibbs Not sure if you want to add a link to "CPU-bound"? Some people might find it interesting to know the difference between CPU and IO bound :) 

http://yaoyao.codes/os/2017/03/20/cpu-bound-vs-io-bound

Reply:
That's a good idea. Done.